src/r_sky.c

```diff
@@ -42,255 +44,361 @@ boolean stretchsky;
     }
-    PrepareFirePixels(fire);
 }
 
-byte *R_GetFireColumn(int col)
+void R_InitFireSky(sky_t *sky)
 {
-    while (col < 0)
+    int texnum = sky->background.texture;
+    const texture_t *tex = textures[texnum];
+    size_t size = tex->width * tex->height;
+    int arr_size = array_size(sky->palette);
+    sky->fire = Z_Calloc(1, size, PU_STATIC, NULL);
+
+    for (int i = 0; i < tex->width; i++)
+    {
+        sky->fire[(tex->height - 1) * tex->width + i] = arr_size - 1;
+    }
+
+    for (int i = 0; i < 64; i++)
     {
-        col += FIRE_WIDTH;
+        UpdateFireSky(sky);
     }
-    col %= FIRE_WIDTH;
-    return &fire_pixels[col * FIRE_HEIGHT];
 }
 
-static void InitSky(void)
+void R_InitSkyMap(void)
 {
-    static skydefs_t *skydefs;
-
     static boolean run_once = true;
     if (run_once)
     {
         skydefs = R_ParseSkyDefs();
         run_once = false;
     }
 
-    if (!skydefs)
+    if (skydefs)
     {
-        return;
-    }
+        flatmap_t *flatmap = NULL;
+        array_foreach(flatmap, skydefs->flatmapping)
+        {
+            int flatnum = R_FlatNumForName(flatmap->flat);
+            int skytex = R_TextureNumForName(flatmap->sky);
+            skyindex_t skyindex = R_AddLevelsky(skytex);
 
-    flatmap_t *flatmap = NULL;
-    array_foreach(flatmap, skydefs->flatmapping)
-    {
-        int flatnum = R_FlatNumForName(flatmap->flat);
-        int skytex = R_TextureNumForName(flatmap->sky);
+            for (int i = 0; i < numsectors; i++)
+            {
+                if (sectors[i].floorpic == flatnum)
+                {
+                    sectors[i].floorpic = skyflatnum;
+                    sectors[i].floorsky = skyindex | PL_SKYFLAT;
+                }
+                if (sectors[i].ceilingpic == flatnum)
+                {
+                    sectors[i].ceilingpic = skyflatnum;
+                    sectors[i].ceilingsky = skyindex | PL_SKYFLAT;
+                }
+            }
+        }
 
-        for (int i = 0; i < numsectors; i++)
+        sky_t *sky;
+        int fg_count = 0;
+        array_foreach(sky, levelskies)
         {
-            if (sectors[i].floorpic == flatnum)
+            if (sky->type == SkyType_WithForeground)
             {
-                sectors[i].floorpic = skyflatnum;
-                sectors[i].floorsky = skytex | PL_FLATMAPPING;
-                R_GetSkyColor(skytex);
+                fg_count++;
             }
-            if (sectors[i].ceilingpic == flatnum)
+        }
+
+        if (fg_count + array_size(levelskies) >= array_capacity(levelskies))
+        {
+            array_grow(levelskies, fg_count);
+        }
+
+        array_foreach(sky, levelskies)
+        {
+            if (sky->type == SkyType_WithForeground)
             {
-                sectors[i].ceilingpic = skyflatnum;
-                sectors[i].ceilingsky = skytex | PL_FLATMAPPING;
-                R_GetSkyColor(skytex);
+                int index = R_AddLevelsky(sky->foreground.texture);
```

Comment:
I have introduced the `skyindex_t` type for `int` type variables with the special measning "level sky index". Please use this consistently or remove it entirely.